cdc/api/v2/safepoint.go

```diff
@@ -0,0 +1,225 @@
+		return safepoint, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		var msg []byte
+		msg, err = io.ReadAll(resp.Body)
+		if err != nil {
+			return safepoint, err
+		}
+		return safepoint, fmt.Errorf("failed to get response: \"[%d] %s", resp.StatusCode, msg)
+	}
+
+	content, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return safepoint, err
+	}
+	if err := json.Unmarshal(content, &safepoint); err != nil {
+		return safepoint, err
+	}
+	for _, s := range safepoint.ServiceGCSafepoints {
+		s.ExpiredTime = time.Unix(s.ExpiredAt, 0).Format(time.RFC3339Nano)
+	}
+	return safepoint, nil
+}
+
+func (h *OpenAPIV2) getServiceID(serviceIDSuffix string) string {
+	if serviceIDSuffix == "" {
+		serviceIDSuffix = defaultServiceIDSuffix
+	}
+	tag := "-" + serviceIDSuffix
+	return h.capture.GetEtcdClient().GetEnsureGCServiceID(tag)
+}
+
+func (h *OpenAPIV2) getPDSafepoint() (*SafePoint, error) {
+	up, err := getCaptureDefaultUpstream(h.capture)
+	if err != nil {
+		return nil, err
+	}
+	return h.helpers.getPDSafepoint(up.PdEndpoints)
+}
+
+// querySafePoint request and returns safepoints from PD
+// @Summary Get the safepoint information of a TiCDC node
+// @Description This API is a synchronous interface. If the request is successful,
+// the safepoint information of the corresponding pd is returned.
+//
+// @Tags common,v2
+// @Produce json
+// @Success 200 {object} SafePoint
+// @Failure 500,400 {object} model.HTTPError
+// @Router	/api/v2/safepoint [get]
+func (h *OpenAPIV2) querySafePoint(c *gin.Context) {
+	safePoint, err := h.getPDSafepoint()
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.IndentedJSON(http.StatusOK, safePoint)
+}
+
+// setSafePoint request and returns safepoints from PD
+// @Summary Set the safepoint information of a TiCDC node
+// @Description This API is a synchronous interface. If the request is successful,
+// the safepoint information of the corresponding pd is returned.
+//
+// @Tags common,v2
+// @Accept json
+// @Produce json
+// @Success 200 {object} SafePoint
+// @Failure 500,400 {object} model.HTTPError
+// @Router	/api/v2/safepoint [post]
+func (h *OpenAPIV2) setSafePoint(c *gin.Context) {
+	ctx := c.Request.Context()
+
+	safePointConfig := &SafePointConfig{}
+	if err := c.BindJSON(safePointConfig); err != nil {
+		_ = c.Error(cerror.WrapError(cerror.ErrAPIInvalidParam, err))
+		return
+	}
+	if safePointConfig.TTL <= 0 {
+		_ = c.Error(cerror.WrapError(cerror.ErrAPIInvalidParam, errors.New("can't set ttl<=0")))
+		return
+	}
+	serviceID := h.getServiceID(safePointConfig.ServiceIDSuffix)
+
+	resp := &SafePoint{}
+	err := h.withUpstreamConfig(ctx, &UpstreamConfig{},
+		func(ctx context.Context, client pd.Client) error {
+			minServiceSafePoint, err := client.UpdateServiceGCSafePoint(ctx, serviceID, safePointConfig.TTL, safePointConfig.StartTs)
+			if err != nil {
+				return err
+			}
+			if minServiceSafePoint > safePointConfig.StartTs {
+				// use minServiceSafePoint to set
```

Comment:
I think we should return an ErrCliInvalidServiceSafePoint error here and not try to set a different safepoint silently.

Reply:
This command is used to prevent GC. It is inconvenient for users to specify a proper safepoint.

Reply:
User can simple use one changefeed's checkpointTs as a safepoint.

Reply:
How about `minServiceSafePoint > safePointConfig.StartTs && safePointConfig.StartTs == 0`? If users don't specify a safepoint, the minServiceSafePoint will be used. The others will return an error.  We define [defaultServiceIDSuffix](https://github.com/pingcap/tiflow/pull/11471/files#diff-0d2d4307b29058ba0f8afd647659b99f6fa28400848ff5796851b556d238dbfeR36) to let users can't update original serviceSafePoints.